src/riscv32i.c

```diff
@@ -119,16 +119,27 @@ static void riscv32i_srli_srai(riscv32_vm_state_t *vm, const uint32_t instructio
     // Perform right arithmetical/logical bitshift on rs1 by imm, store into rds
     uint32_t rds = cut_bits(instruction, 7, 5);
     uint32_t rs1 = cut_bits(instruction, 15, 5);
-    uint32_t shamt = cut_bits(instruction, 20, 5);
-    uint32_t src_reg = riscv32i_read_register_s(vm, rs1);
-
-    uint32_t funct7 = cut_bits(instruction, 25, 7);
+    uint32_t shamt = cut_bits(instruction, 20, XLEN_BIT(vm));
+    uint32_t funct7 = cut_bits(instruction, 20 + XLEN_BIT(vm), 12 - XLEN_BIT(vm));
+
+    reg_t src_reg = riscv32i_read_register_u(vm, rs1);
+
+    size_t extend_from = XLEN(vm);
+    for (size_t i = 5; i < XLEN_BIT(vm); ++i)
+    {
+	    if (!is_bit_set(shamt, i))
+	    {
+		    extend_from = 1 << i;
+		    src_reg &= gen_mask(extend_from);
+		    break;
+	    }
+    }
```

Comment:
This should not be done here, separate instruction implementations with separate dispatch tables are going to be used for RV64. New srli/srai/slli are 3-4 times slower than the original implementation, and this is likely to be redundant. I suggest reverting this change.